Turned the backend into a QML plugin.

Druage opened the pull request:
This now turns the project into a QML plugin. Statically linking to this backend has been removed, but canquickly added by anyone. ( It's not hard. )

Commit:
Added compilation step to create qml plugins.

Commit:
The backend now completely builds as a qml plugin.

Commit:
Sectioned the .pro file.

Druage commented:
We cannot merge this PR, until the "feature-importer-mt" branch of Phoenix, has been merged into its "master" branch because this "qmlplugin" branch was based off of the "feature-importer-mt" branch.

Druage commented:
I forgot about how the frontend is synced to a specific commit in master, so this won't affect that.

athairus commented on backend.pro:
```diff
@@ -1,78 +1,91 @@
+#################
 
-    # Include libraries
+    # Include platform specific libraries
```
I think it looked neater before...

athairus commented on backend.pro:
```diff
@@ -1,78 +1,91 @@
-## Extra targets
-##
+#################
+# Extra targets #
```
You And I Are Different When It Comes To Section Headers, I don't like to use Title Case (and for some reason you didn't use it here lol)

Commit:
Removed backend_plugins.pro, it wasn't needed.

athairus commented on backendplugin.cpp:
```diff
@@ -0,0 +1,32 @@
+#include "backendplugin.h"
```
Don't include anything in .cpp files other than their corresponding header!

athairus commented on backendcommon.h:
```diff
@@ -14,10 +14,6 @@
 
 #include <math.h>
 
-#include "samplerate.h"
```
Why were these removed? Are you not using the precompiled header anymore?

athairus commented on backendplugin.h:
```diff
@@ -0,0 +1,14 @@
+#ifndef COREPLUGIN_H
+#define COREPLUGIN_H
+
+#include <QQmlExtensionPlugin>
```
Use the common header

athairus commented on consumer/audiooutput.cpp:
```diff
@@ -1,5 +1,5 @@
-
 #include "audiooutput.h"
+#include "logging.h"
```
Don't include anything in the .cpp file besides the class's header

athairus commented on consumer/audiooutput.h:
```diff
@@ -5,11 +5,9 @@
 #include "consumer.h"
 #include "controllable.h"
-
+#include <samplerate.h>
```
Use precompiled header, separate from our own headers (put `#include "backendcommon.h"` above `#include "consumer.h"` with a blank line)

athairus commented on consumer/videooutput.cpp:
```diff
@@ -32,6 +32,9 @@ VideoOutput::~VideoOutput() {
 
 void VideoOutput::consumerFormat( ProducerFormat format ) {
+
+    qDebug() << "CONSUMER FORMAT START";
```
Didn't you make a big stink a long time ago about making sure to tag debug messages? :)

Druage replied:
Yes, this was meant to be deleted.

athairus commented on core/core.cpp:
```diff
@@ -1,5 +1,8 @@
 #include "core.h"
 
+#include "logging.h"
```
Don't include anything in the .cpp file besides the class's header

athairus commented on core/core.h:
```diff
@@ -3,8 +3,6 @@
 
 #include "backendcommon.h"
 
-#include "logging.h"
```
Good, this should go in backendcommon anyway

athairus commented on core/corecontrolproxy.cpp:
```diff
@@ -1,5 +1,8 @@
 #include "corecontrolproxy.h"
 
+#include <type_traits>
+#include <QQmlApplicationEngine>
```
Don't include anything in the .cpp file besides the class's header

athairus commented on core/corecontrolproxy.h:
```diff
@@ -59,10 +59,10 @@ class CoreControlProxy : public QObject {
 
         // Needed by: "libretro"
-        Q_PROPERTY( VideoOutput *videoOutput READ getVideoOutput WRITE setVideoOutput NOTIFY videoOutputChanged )
+        Q_PROPERTY( QObject *videoOutput READ getVideoOutput WRITE setVideoOutput NOTIFY videoOutputChanged )
```
Why are these changed?

athairus commented on input/inputmanager.cpp:
```diff
@@ -1,18 +1,23 @@
 #include "inputmanager.h"
 
+#include "joystick.h"
```
Don't include anything in the .cpp file besides the class's header

athairus commented on input/inputmanager.h:
```diff
@@ -6,13 +6,13 @@
-
 #include "logging.h"
 
+class InputDevice;
```
I can see you doing your forward declarations here. For consistency's sake, we should use my style.

athairus commented on input/sdleventloop.cpp:
```diff
@@ -1,10 +1,13 @@
 #include "sdleventloop.h"
-
 #include "logging.h"
+#include <SDL.h>
```
Don't include anything in the .cpp file besides the class's header

athairus commented on input/sdleventloop.h:
```diff
@@ -1,11 +1,13 @@
+class Joystick;
 
 // The SDLEventLoop's job is to poll for button states and to emit signals as devices are connected or disconnected.
+#include <QObject>
```
Use the precompiled header

athairus commented on util/logging.h:
```diff
@@ -2,6 +2,7 @@
 #define LOGGING_H
 
 #include "backendcommon.h"
+#include <QDebug>
```
Should be covered by precompiled header

athairus commented on backend.pro:
```diff
@@ -96,39 +109,42 @@
+        backendplugin.cpp
+____________________________________________________________________
+####################
+# Linkers Settings #
```
Linker

Commit:
Moved header includes to only where they are needed.
The backendcommon.h file now only contains typedef's and the rest of the files that
depended upon it now include the headers directly, but only exactly when needed.
I tried to use forward declares if I could, and included headers if nothing else could
be done. Most includes were moved to the .cpp files.

athairus committed:
Misc cleanup

- Removed the make install target for qmldir

Commit:
Removed debug statement, reverted dynamic cast in CCProxy.
